File: internal/events/events.go
// Package events defines the durable audit record that explains every
// lifecycle decision. Events never contain prompts, generated content,
// request bodies, or credentials.
package events

import (
	"time"

	"github.com/sentania-labs/benchwarmer/internal/policy"
	"github.com/sentania-labs/benchwarmer/internal/state"
)

// Type classifies an event. Stable strings: stored, exported, and shown.
type Type string

// Event types.
const (
	ServiceStarted  Type = "service_started"
	ServiceStopping Type = "service_stopping"
	StateChanged    Type = "state_changed"
	// DecisionChanged records a change of winning rule without a state
	// change (e.g. cooldown reason changes from game to soft contention).
	DecisionChanged Type = "decision_changed"

	LoadStarted   Type = "load_started"
	LoadCompleted Type = "load_completed"
	LoadFailed    Type = "load_failed"

	DrainStarted    Type = "drain_started"
	DrainCompleted  Type = "drain_completed" // requests finished within grace
	GraceShortened  Type = "grace_shortened"
	Preempted       Type = "preempted" // runtime force-terminated with requests active
	RuntimeStopped  Type = "runtime_stopped"
	RuntimeCrashed  Type = "runtime_crashed"
	KillFailed      Type = "kill_failed"
	VRAMReleased    Type = "vram_released"
	VRAMNotReleased Type = "vram_release_timeout"
	OrphanKilled    Type = "orphan_killed"

	CooldownStarted    Type = "cooldown_started"
	CooldownReset      Type = "cooldown_reset"
	SuppressionStarted Type = "suppression_started"
	RecoveryScheduled  Type = "recovery_scheduled"

	ModeChanged     Type = "mode_changed"
	ModeExpired     Type = "mode_expired"
	ConfigChanged   Type = "config_changed"
	ConfigRejected  Type = "config_rejected"
	ConfigRecovered Type = "config_recovered"

	TelemetryLost      Type = "telemetry_lost"
	TelemetryRecovered Type = "telemetry_recovered"
	TelemetryDegraded  Type = "telemetry_degraded"

	PowerSuspend      Type = "power_suspend"
	PowerResume       Type = "power_resume"
	SessionChange     Type = "session_change"
	AgentConnected    Type = "agent_connected"
	AgentDisconnected Type = "agent_disconnected"
	DeviceLost        Type = "device_lost"

	// RequestsRejected aggregates 503s during one unavailable period rather
	// than one event per request.
	RequestsRejected Type = "requests_rejected"
	// RequestForceClosed records a request cut off by preemption or the
	// maximum request duration (ID and duration only).
	RequestForceClosed Type = "request_force_closed"
)

// Telemetry is the compact telemetry snapshot stored with an event.
type Telemetry struct {
	Confidence      policy.Confidence `json:"confidence"`
	TotalUtilPct    float64           `json:"total_util_pct"`
	OwnUtilPct      float64           `json:"own_util_pct"`
	ExternalUtilPct float64           `json:"external_util_pct"`
	VRAMUsedMiB     int               `json:"vram_used_mib"`
	VRAMFreeMiB     int               `json:"vram_free_mib"`
	OwnVRAMMiB      int               `json:"own_vram_mib"`
	ExternalVRAMMiB int               `json:"external_vram_mib"`
	TemperatureC    *float64          `json:"temperature_c,omitempty"`
}

// TelemetryFrom extracts the stored subset from policy facts.
func TelemetryFrom(g policy.GPUFacts) *Telemetry {
	return &Telemetry{
		Confidence: g.Confidence, TotalUtilPct: g.TotalUtilPct, OwnUtilPct: g.OwnUtilPct,
		ExternalUtilPct: g.ExternalUtilPct, VRAMUsedMiB: g.VRAMUsedMiB, VRAMFreeMiB: g.VRAMFreeMiB,
		OwnVRAMMiB: g.OwnVRAMMiB, ExternalVRAMMiB: g.ExternalVRAMMiB, TemperatureC: g.TemperatureC,
	}
}

// Event is one audit record.
type Event struct {
	ID   int64     `json:"id"` // assigned by the store
	Time time.Time `json:"time"`
	Type Type      `json:"type"`

	PrevState state.State     `json:"prev_state,omitempty"`
	State     state.State     `json:"state,omitempty"`
	Condition state.Condition `json:"condition,omitempty"`

	Profile  string            `json:"profile,omitempty"`
	Mode     policy.Mode       `json:"mode,omitempty"`
	Rule     string            `json:"rule,omitempty"`
	Severity policy.Severity   `json:"severity,omitempty"`
	Evidence []policy.Evidence `json:"evidence,omitempty"`

	Telemetry *Telemetry `json:"telemetry,omitempty"`

	RuntimePID int    `json:"runtime_pid,omitempty"`
	RequestID  string `json:"request_id,omitempty"`

	// Message is the human-readable explanation.
	Message string `json:"message"`
	// Data holds type-specific measurements (durations in milliseconds as
	// *_ms, deadlines as RFC 3339). Never prompts or bodies.
	Data map[string]any `json:"data,omitempty"`
}

// FromDecision starts an event populated from a policy decision.
func FromDecision(t Type, now time.Time, d policy.Decision, g policy.GPUFacts) Event {
	return Event{
		Time: now, Type: t, Profile: d.Profile, Mode: d.Mode, Rule: d.Rule,
		Severity: d.Severity, Evidence: d.Evidence, Telemetry: TelemetryFrom(g), Message: d.Reason,
	}
}

// Sink receives events. Implementations must not block the controller for
// long; the store buffers and writes asynchronously.
type Sink interface {
	Emit(Event)
}

// SinkFunc adapts a function to Sink.
type SinkFunc func(Event)

// Emit implements Sink.
func (f SinkFunc) Emit(e Event) { f(e) }

File: internal/webui/static/js/events.js
// Events page: GET api/v1/events, newest first, paged with before_id.
import { h, replace } from "./dom.js";
import { get, signIn, ApiError } from "./api.js";
import { clockTime, mib, pct, parseTime, titleize } from "./format.js";
import { evidenceTable } from "./dashboard.js";

const PAGE = 50;

// Mirrors internal/events Type constants.
const TYPES = [
  "service_started", "service_stopping", "state_changed", "decision_changed",
  "load_started", "load_completed", "load_failed",
  "drain_started", "drain_completed", "grace_shortened", "preempted", "runtime_stopped",
  "runtime_crashed", "kill_failed", "vram_released", "vram_release_timeout", "orphan_killed",
  "cooldown_started", "cooldown_reset", "suppression_started", "recovery_scheduled",
  "mode_changed", "mode_expired", "config_changed", "config_rejected", "config_recovered",
  "telemetry_lost", "telemetry_recovered", "telemetry_degraded",
  "power_suspend", "power_resume", "session_change", "agent_connected", "agent_disconnected", "device_lost",
  "requests_rejected", "request_force_closed",
];

let state = null;

export function mount(main) {
  const filter = h("select", { id: "event-type", onchange: () => load(true) },
    h("option", { value: "" }, "All types"),
    TYPES.map((t) => h("option", { value: t }, titleize(t))));
  const refresh = h("button", { type: "button", onclick: () => load(true) }, "Refresh");
  const list = h("ol", { class: "events", "aria-label": "Events, newest first" });
  const more = h("button", { type: "button", hidden: true, onclick: () => load(false) }, "Load older events");
  const status = h("p", { class: "muted", role: "status" });
  state = { filter, list, more, status, before: 0, loading: false };
  replace(main,
    h("h1", null, "Events"),
    h("div", { class: "toolbar" },
      h("div", { class: "field inline" }, h("label", { for: "event-type" }, "Type"), filter), refresh),
    status, list, more);
  load(true);
  return { title: "Events", unmount: () => { state = null; } };
}

async function load(reset) {
  const s = state;
  if (!s || s.loading) return;
  s.loading = true;
  if (reset) {
    s.before = 0;
    replace(s.list);
  }
  s.status.textContent = "Loading...";
  const q = new URLSearchParams({ limit: String(PAGE) });
  if (s.filter.value) q.set("type", s.filter.value);
  if (s.before) q.set("before_id", String(s.before));
  try {
    const res = await get("api/v1/events?" + q.toString(), { interactive: false });
    if (state !== s) return;
    const evs = res.events || [];
    const now = new Date();
    for (const e of evs) s.list.append(row(e, now));
    s.before = res.next_before_id || 0;
    s.more.hidden = !s.before;
    s.status.textContent = s.list.children.length ? "" : "No events.";
  } catch (e) {
    if (state !== s) return;
    if (e instanceof ApiError && e.status === 401) {
      replace(s.status, "Reading events needs the management token. ",
        h("button", { type: "button", onclick: async () => { if (await signIn()) load(true); } }, "Enter token"));
    } else {
      s.status.textContent = "Could not load events: " + e.message;
    }
  } finally {
    s.loading = false;
  }
}

function telemetryTable(t) {
  const rows = [
    ["Confidence", t.confidence],
    ["Total utilization", pct(t.total_util_pct)],
    ["Benchwarmer utilization", pct(t.own_util_pct)],
    ["External utilization", pct(t.external_util_pct)],
    ["VRAM used", mib(t.vram_used_mib)],
    ["VRAM free", mib(t.vram_free_mib)],
    ["Benchwarmer VRAM", mib(t.own_vram_mib)],
    ["External VRAM", mib(t.external_vram_mib)],
    t.temperature_c != null && ["Temperature", Math.round(t.temperature_c) + " \u00b0C"],
  ].filter(Boolean);
  return h("dl", { class: "kv" }, rows.map(([k, v]) => [h("dt", null, k), h("dd", null, v)]));
}

// dataValue renders RFC 3339 strings in Data as local clock times.
function dataValue(v) {
  if (typeof v === "string" && /^\d{4}-\d\d-\d\dT\d\d:\d\d/.test(v)) {
    const d = parseTime(v);
    if (d) return clockTime(d, null, { seconds: true });
  }
  return typeof v === "object" ? JSON.stringify(v) : String(v);
}

function row(e, now) {
  const t = parseTime(e.time);
  const change = e.prev_state || e.state
    ? (e.prev_state || "?") + " \u2192 " + (e.state || "?") : "";
  const sev = e.severity || "info";
  const details = [];
  if (e.evidence && e.evidence.length) details.push(h("h3", null, "Evidence"), evidenceTable(e.evidence));
  if (e.telemetry) details.push(h("h3", null, "Telemetry"), telemetryTable(e.telemetry));
  if (e.data && Object.keys(e.data).length) {
    details.push(h("h3", null, "Data"), h("dl", { class: "kv" },
      Object.entries(e.data).map(([k, v]) => [h("dt", null, k), h("dd", null, dataValue(v))])));
  }
  const meta = [
    e.condition && "Condition: " + e.condition,
    e.profile && "Profile: " + e.profile,
    e.mode && "Mode: " + e.mode,
    e.runtime_pid && "PID: " + e.runtime_pid,
    e.request_id && "Request: " + e.request_id,
    "Event #" + e.id,
  ].filter(Boolean).join(" \u00b7 ");
  details.push(h("p", { class: "muted small" }, meta));
  return h("li", { class: "event sev-" + sev },
    h("details", null,
      h("summary", null,
        h("time", { datetime: e.time }, t ? clockTime(t, now, { seconds: true }) : ""),
        h("span", { class: "etype" }, titleize(e.type)),
        change && h("span", { class: "change" }, change),
        e.rule && h("code", { class: "rule" }, e.rule),
        h("span", { class: "msg" }, e.message || "")),
      h("div", { class: "event-body" }, details)));
}
